File: SDK/Linux/TmLinux/FirmwareWorker.cpp
/******************************************************************
 * Project: TmSDK
 * File: FirmwareWorker.cpp
 *
 * Description: This file contains the following implementations:
 * - Thread class for updating camera firmware.
 *
 * Version: 1.0.0
 * Copyright 2024. Thermoeye Inc. All rights reserved.
 *
 * History:
 *      2024-08-19: Initial version.
 ****************************************************************/
#include <QThread>
#include <QMetaType>
#include <iostream>
#include "FirmwareWorker.h"

Q_DECLARE_METATYPE(QString)

/*
* Constructor of FirmwareWorker class
* parameter:
*   camera: Pointer to the TmCamera
*   parent: Pointer to the parent QObject
*/
FirmwareWorker::FirmwareWorker(TmSDK::TmCamera* camera, QObject* parent)
    : QThread(parent), stopFlag(false), pTmCamera(camera)
{
    qRegisterMetaType<QString>("QString");
}

/**
 * Executes the firmware update thread.
 */
void FirmwareWorker::run()
{
    int percent = 0;

    if (pTmCamera != nullptr && pTmCamera->IsOpen()) {
        while (percent < 100 && !stopFlag)
        {
            percent = pTmCamera->pTmControl->UpdateFirmware();
            if (percent >= 0)
            {
                emit ProgressChanged(percent);
            }
            else
            {
                emit WorkCompleted(false, "Error during download firmware.");
                return;
            }
        }

        if (stopFlag)
        {
            emit WorkCompleted(false, "Update cancelled.");
        }
        else
        {
            emit WorkCompleted(true, "Update completed.");
        }
    }
}

/**
 * Stop the firmware update thread.
 */
void FirmwareWorker::cancel()
{
    stopFlag = true;
}

File: SDK/Linux/examples/TmPython/FirmwareWorker.py
#################################################################
# Company: Thermoeye, Inc
# Project: TmSDK
# File: FirmwareWorker.py
# Creation date: 2024-08-19
# Version: 1.0.0
#
# Description: This file contains the following implementations:
#  - Thread class for updating camera firmware.
#
# History: 2024-08-19: Initial version.
#
#################################################################
from PyQt5.QtCore import QThread, pyqtSignal
from PyQt5.QtCore import QObject

class FirmwareWorker(QThread):
    progressChanged = pyqtSignal(int)
    workCompleted = pyqtSignal(bool, str)

    def __init__(self, parent):
        super().__init__();
        self.stopFlag = False
        self.parent = parent
        self.tmCamera = parent.camera.tmCamera

    def run(self):
        percent = 0
        if self.tmCamera is not None:
            while percent < 100 and not self.stopFlag:
                percent = self.tmCamera.tmControl.update_firmware()
                if percent >= 0:
                    self.progressChanged.emit(percent)
                else:
                    self.workCompleted.emit(False, "Error during download firmware.")
                    return

            if self.stopFlag:
                self.workCompleted.emit(False, "Update cancelled.")
            else:
                self.workCompleted.emit(True, "Update completed.")

    def stop(self):
        self.stopFlag = True
